src/Npgsql/TypeHandlers/DateTimeHandlers/TimeTzHandler.cs

```diff
@@ -41,10 +41,21 @@ class TimeTzHandler : SimpleTypeHandler<DateTimeOffset>, ISimpleTypeHandler<Date
 
         public override DateTimeOffset Read(ReadBuffer buf, int len, FieldDescription fieldDescription = null)
         {
-            // Adjust from 1 microsecond to 100ns. Time zone (in seconds) is inverted.
-            var ticks = buf.ReadInt64() * 10;
-            var offset = new TimeSpan(0, 0, -buf.ReadInt32());
-            return new DateTimeOffset(ticks, offset);
+            // Adjust from 1 microsecond to 100ns.
+            var timeTicks = buf.ReadInt64() * 10L;
+            // Invert time zone, adjust from seconds to 100ns.
+            var offsetTicks = buf.ReadInt32() * -10000000L;
+
+            var offset = new TimeSpan(offsetTicks);
+            if (timeTicks - offsetTicks < 0L)
```

Comment:
I might be misreading your code, but doesn't this add a day only when the offset-adjusted time is negative? In other words, this seems like it would usually produce DateTimeOffsets with date 0001-01-01, but occasionally it would produce instances with date 0001-01-02...

This is the part that's bothering me, I'd rather it produced DateTimeOffsets with the same date. This can be 0001-01-02, or maybe 2000-01-01 (which looks slightly less arbitrary, even though it is).

Reply:
It's okay, because a `DateTimeOffset` will return the coerced value for any of its properties except `LocalDateTime`, `UtcDateTime` and `UtcTicks`.

Reply:
@YohDeadfall, I didn't understand your comment, maybe we're talking about different things...

My problem is that in this PR, the following test fails:

```c#
using (var conn = OpenConnection())
using (var cmd = new NpgsqlCommand("SELECT TIME WITH TIME ZONE '01:00:00+02', TIME WITH TIME ZONE '05:00:00+02'", conn))
using (var reader = cmd.ExecuteReader())
{
    reader.Read();
    Assert.That(reader.GetFieldValue<DateTimeOffset>(0).Date == reader.GetFieldValue<DateTimeOffset>(1).Date);
}
```

In other words, the day component returned isn't always the same - which I find problematic. The fixup proposed by this PR only occurs when `timeTicks - offsetTicks < 0L`, so in some cases day is 1, and in others it's 2. Maybe you fixed this in your own PR.

Regarding which branch to submit against, I think this is something we want in 4.0.0, so you can submit against dev. I haven't created a branch for 4.0.0-rtm (as opposed to 4.1.0) - I think that for the next few weeks we can concentrate on 4.0 work which can happen in dev, and only after RTM we can branch off a hotfix/4.0.1 at which point dev becomes 4.1 (similar to how we currently have hotfix/3.2.8).

(In fact, I think it'll be healthy to focus on 4.0 for a few weeks even after RTM, until 4.0 stabilizes, and only start big work for 4.1 a bit later).

Let me know what you think - the above is just my opinion. I'll wait to see your PR which replaces this one.

Reply:
Sorry for the very delayed answer. The issue with the date component changing is, well, not something we can do anything about. Either we set the UTC representation of the DateTimeOffset to a fixed date (what my PR effectively does), which means when representing it in the correct time zone it won't be stable. Or, without the fix, the UTC representation becomes unstable (the day being either 1 or 0, which triggered the error). I figured the date part would be discarded anyways, so I decided on fixing the UTC day component to 1 (by adding 1 to it when it would be 0 otherwise). At the end I don't care about the date part anyways, so feel free to use a different approach, but we won't really solve the problem either way.

Reply:
@roji That's also a valid approach. The difference is that in my PR, the day would only change in situations where earlier Npgsql versions would crash, therefore not introducing changes to already working code. But in the end, both solutions would work just fine.

Reply:
You are both right at some point, because @roji's solution is free from the problem demonstrated in https://github.com/npgsql/npgsql/pull/1662%23discussion_r187327524. But @nkreipke's solution doesn't require changes in user code. If a new version of Npgsql is incompatible with previous, the solution mentioned by @roji will be better.

/cc @dodyg, @QuanTran91

Reply:
@YohDeadfall and @nkreipke - you're right, what I'm proposing is a breaking change (even though it's a pretty minor one...). That's also why this is the perfect time to do it - for 4.0.0 which introduces several other breaking changes as well. Since we have this opportunity, I prefer if we implemented the best/most logical behavior even if means breakage.

(of course, assuming that you agree that my proposal is the best behavior)

Reply:
@roji In the end, both solutions seem kind of hacky, but I guess that's the best we can do with .NET builtin types. If we introduce breaking changes anyway, I'd just drop support for DateTimeOffset in favor of OffsetDateTime :wink: 

Joking aside, I'd go for your proposal, it's definitely cleaner. We should definitely warn users about this change though.

Reply:
@nkreipke I totally agree, the .NET builtin date/time types simply aren't very good, which is why in 4.0 NodaTime will be the recommended way to do date/time. Although in this case I have to admit that I also don't really understand the usefulness of PostgreSQL `timetz`; it's quite a weird type, especially considering that there's no actual timestamp with timezone (i.e. a real equivalent to .NET DateTimeOffset).

Anyway...

Reply:
I think that's been there for a while, not sure what the plan is for evolving the BCL date/time types...

In any case I don't think there's ever going to be a BCL type that's actually going to correspond to PostgreSQL timetz (and rightfully so)...